[IntelliJ][ReportedByUser] Uncaught Exception  java.lang.IllegalStateException: Fatal Exception thrown on Scheduler.Worker thread.

IntelliJ build version: 2019.3.3 IU-193.6494.35
OS: Windows 10
JDK: JetBrains s.r.o 11.0.5
Plugin version: 3.33.1-2019.3
Additional Info: None
Parent component: 
```
com.intellij.openapi.wm.impl.IdeRootPane[,8,8,1921x1040,layout=com.intellij.openapi.wm.impl.IdeRootPane$MyRootLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=449,maximumSize=,minimumSize=,preferredSize=]
```
Error message: 
```
IdeaLoggingEvent[message=null, throwable=java.lang.IllegalStateException: Fatal Exception thrown on Scheduler.Worker thread.
	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:59)
Caused by: rx.exceptions.OnErrorFailedException: Error occurred when trying to propagate error to Observer.onError
Caused by: rx.exceptions.CompositeException: 2 exceptions occurred. 
Caused by: rx.exceptions.CompositeException$CompositeExceptionCausalChain: Chain of Causes for CompositeException In Order Received =>
Caused by: java.lang.Error: The process has been terminated
	at com.microsoft.intellij.runner.RunProcessHandler.println(RunProcessHandler.java:86)
	at com.microsoft.intellij.runner.RunProcessHandler.setText(RunProcessHandler.java:120)
	at com.microsoft.azuretools.utils.WebAppUtils.deployArtifactsToAppService(WebAppUtils.java:321)
	at com.micro

--- a/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java
+++ b/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java
@@ -71,7 +71,7 @@ public abstract class AzureRunProfileState<T> implements RunProfileState {
                     return this.executeSteps(processHandler, telemetryMap);
                 } finally {
                     // Once the operation done, whether success or not, `setText` should not throw new exception
-                    processHandler.setProcessTerminatedHandlerForSetText(RunProcessHandler.DO_NOTHING);
+                    processHandler.setProcessTerminatedHandler(RunProcessHandler.DO_NOTHING);
                 }
             }).subscribeOn(SchedulerProviderFactory.getInstance().getSchedulerProvider().io()).subscribe(
                 (res) -> {

--- a/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java
+++ b/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/AzureRunProfileState.java
@@ -63,29 +63,34 @@ public abstract class AzureRunProfileState<T> implements RunProfileState {
         final Operation operation = createOperation();
         Observable.fromCallable(
             () -> {
-                if (operation != null) {
-                    operation.start();
-                    EventUtil.logEvent(EventType.info, operation, telemetryMap);
+                try {
+                    if (operation != null) {
+                        operation.start();
+                        EventUtil.logEvent(EventType.info, operation, telemetryMap);
+                    }
+                    return this.executeSteps(processHandler, telemetryMap);
+                } finally {
+                    // Once the operation done, whether success or not, `setText` should not throw new exception
+                    processHandler.setProcessTerminatedHandlerForSetText(RunProcessHandler.DO_NOTHING);
                 }
-                return this.executeSteps(processHandler, telemetryMap);
             }).subscribeOn(SchedulerProviderFactory.getInstance().getSchedulerProvider().io()).subscribe(
-            (res) -> {
-                if (operation != null) {
-                    operation.complete();
-                }
-                this.sendTelemetry(telemetryMap, true, null);
-                this.onSuccess(res, processHandler);
-            },
-            (err) -> {
-                err.printStackTrace();
-                if (operation != null) {
-                    EventUtil.logError(operation, ErrorType.userError, new Exception(err.getMessage(), err),
-                        telemetryMap, null);
-                    operation.complete();
-                }
-                this.onFail(err, processHandler);
-                this.sendTelemetry(telemetryMap, false, err.getMessage());
-            });
+                (res) -> {
+                    if (operation != null) {
+                        operation.complete();
+                    }
+                    this.sendTelemetry(telemetryMap, true, null);
+                    this.onSuccess(res, processHandler);
+                },
+                (err) -> {
+                    err.printStackTrace();
+                    if (operation != null) {
+                        EventUtil.logError(operation, ErrorType.userError, new Exception(err.getMessage(), err),
+                                           telemetryMap, null);
+                        operation.complete();
+                    }
+                    this.onFail(err, processHandler);
+                    this.sendTelemetry(telemetryMap, false, err.getMessage());
+                });
         return new DefaultExecutionResult(consoleView, processHandler);
     }
 
@@ -107,7 +112,8 @@ public abstract class AzureRunProfileState<T> implements RunProfileState {
         }
     }
 
-    protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap){}
+    protected void updateTelemetryMap(@NotNull Map<String, String> telemetryMap) {
+    }
 
     private void sendTelemetry(@NotNull Map<String, String> telemetryMap, boolean success, @Nullable String errorMsg) {
         updateTelemetryMap(telemetryMap);

--- a/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/RunProcessHandler.java
+++ b/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/RunProcessHandler.java
@@ -28,7 +28,6 @@ import com.intellij.execution.process.ProcessListener;
 import com.intellij.execution.process.ProcessOutputTypes;
 import com.intellij.openapi.util.Key;
 import com.microsoft.azuretools.utils.IProgressIndicator;
-
 import org.jetbrains.annotations.Nullable;
 
 import java.io.OutputStream;
@@ -36,6 +35,14 @@ import java.io.OutputStream;
 public class RunProcessHandler extends ProcessHandler implements IProgressIndicator {
 
     private static final String PROCESS_TERMINATED = "The process has been terminated";
+
+    public static final Runnable THROW_TERMINATED_EXCEPTION = () -> {
+        throw new RuntimeException(PROCESS_TERMINATED);
+    };
+    public static final Runnable DO_NOTHING = () -> {
+    };
+
+    private Runnable processTerminatedHandler;
 
     @Override
     protected void destroyProcessImpl() {
@@ -75,7 +82,7 @@ public class RunProcessHandler extends ProcessHandler implements IProgressIndica
         if (isProcessRunning()) {
             this.notifyTextAvailable(message, type);
         } else {
-            throw new Error(PROCESS_TERMINATED);
+            processTerminatedHandler.run();
         }
     }
 
@@ -88,7 +95,7 @@ public class RunProcessHandler extends ProcessHandler implements IProgressIndica
         if (isProcessRunning()) {
             this.notifyTextAvailable(message + "\n", type);
         } else {
-            throw new Error(PROCESS_TERMINATED);
+            processTerminatedHandler.run();
         }
     }
 
@@ -96,6 +103,7 @@ public class RunProcessHandler extends ProcessHandler implements IProgressIndica
      * Process handler to show the progress message.
      */
     public RunProcessHandler() {
+        processTerminatedHandler = THROW_TERMINATED_EXCEPTION;
     }
 
     public void addDefaultListener() {
@@ -141,5 +149,9 @@ public class RunProcessHandler extends ProcessHandler implements IProgressIndica
     @Override
     public boolean isCanceled() {
         return false;
+    }
+
+    public void setProcessTerminatedHandlerForSetText(Runnable runnable) {
+        this.processTerminatedHandler = runnable;
     }
 }

--- a/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/RunProcessHandler.java
+++ b/PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/RunProcessHandler.java
@@ -151,7 +151,7 @@ public class RunProcessHandler extends ProcessHandler implements IProgressIndica
         return false;
     }
 
-    public void setProcessTerminatedHandlerForSetText(Runnable runnable) {
-        this.processTerminatedHandler = runnable;
+    public void setProcessTerminatedHandler(Runnable runnable) {
+        this.processTerminatedHandler = runnable == null ? DO_NOTHING : runnable;
     }
 }

